internal/keepers/service.go

```diff
@@ -202,17 +183,57 @@ func (s *onDemandUpkeepService) parallelCheck(ctx context.Context, keys []types.
 	wg.Wait()
 	close(done)
 
+	if wResults.Total() == 0 {
+		s.logger.Printf("no network calls were made for this sampling set")
+	} else {
+		s.logger.Printf("worker call success rate: %.2f; failure rate: %.2f; total calls %d", wResults.SuccessRate(), wResults.FailureRate(), wResults.Total())
+	}
+
 	s.logger.Printf("sampling cache hit ratio %d/%d", cacheHits, len(keys))
 
+	// multiple network calls can result in an error while some can be successful
+	// in the case that all workers encounter an error, bubble this up as a hard
+	// failure of the process.
+	if wResults.Total() > 0 && wResults.Total() == wResults.Failure && wResults.LastErr != nil {
+		return samples.Values(), fmt.Errorf("%w: too many errors in parallel worker process; last error provided", wResults.LastErr)
+	}
+
 	return samples.Values(), nil
 }
 
+func (s *onDemandUpkeepService) aggregateWorkerResults(w *sync.WaitGroup, r *workerResults, sa *syncedArray[*types.UpkeepResult], done chan struct{}) {
+	s.logger.Printf("starting service to read worker results")
+
+Outer:
+	for {
+		select {
+		case result := <-s.workers.results:
+			w.Done()
+			if result.Err == nil {
+				r.Success++
```

Comment:
not sure if this requires a mutex now?

Reply:
Yes because the workers add to the list at the same time as the for loop that starts workers.